src/OSPSuite.Core/Domain/Services/ReactionCreator.cs

```diff
@@ -35,67 +31,57 @@ internal class ReactionCreator : IReactionCreator, IVisitor<IContainer>
 
       public bool CreateReaction(ReactionBuilder reactionBuilder, ModelConfiguration modelConfiguration)
       {
-         _reactionBuilder = reactionBuilder;
-         (_model, _simulationBuilder) = modelConfiguration;
-         try
-         {
-            //global container should be created before creating local reaction so that path replacement works
-            var reactionGlobalContainer = createReactionPropertiesContainer();
-            _createdInstance = false;
-            _model.Root.AcceptVisitor(this);
+         var (model, simulationBuilder) = modelConfiguration;
+         var canCreateReaction = canCreateReactionInDef(reactionBuilder);
+         var createLocalReaction = createLocalReactionDef(reactionBuilder, simulationBuilder);
+         var allParentContainersForReaction = model.Root.GetAllContainersAndSelf<IContainer>().Where(canCreateReaction).ToList();
+         if (!allParentContainersForReaction.Any())
+            return false;
 
-            //reaction was not created. No need to create a global container
-            if (!_createdInstance)
-               _model.Root.RemoveChild(reactionGlobalContainer);
+         //First create the global reaction container for this reaction
+         createReactionPropertiesContainer(reactionBuilder, modelConfiguration);
 
-            return _createdInstance;
-         }
-         finally
-         {
-            _reactionBuilder = null;
```

Comment:
all of those globals because of the usage of visitor. Clearly not the right usage. Instead, use link to retrieve the containers and be done with it